catalyst-toolbox-cli/src/cli/recovery/tally/fragments.rs

```diff
@@ -0,0 +1,94 @@
+use std::fs;
+use std::io;
+use std::io::BufReader;
+use std::path::{Path, PathBuf};
+
+use bincode::Options;
+use jormungandr_lib::interfaces::PersistentFragmentLog;
+
+struct FileFragments {
+    reader: BufReader<fs::File>,
+}
+
+struct FileFragmentsIterator {
+    reader: BufReader<fs::File>,
+}
+
+impl FileFragments {
+    pub fn from_path(file_path: PathBuf) -> std::io::Result<Self> {
+        fs::File::open(file_path).map(|file| Self {
+            reader: BufReader::new(file),
+        })
+    }
+}
+
+impl IntoIterator for FileFragments {
+    type Item = PersistentFragmentLog;
+    type IntoIter = FileFragmentsIterator;
+
+    fn into_iter(self) -> Self::IntoIter {
+        FileFragmentsIterator {
+            reader: self.reader,
+        }
+    }
+}
+
+impl Iterator for FileFragmentsIterator {
+    type Item = PersistentFragmentLog;
+
+    fn next(&mut self) -> Option<Self::Item> {
```

Comment:
This is some nice idiomatic Rust here!

One note: I think it is actually fine to have `Result` as `Item`. It is easy to process them in a `for` loop and if you use combinators + `collect` then it is just `.collect::<Result<Vec<_>, _>>`. This gives us proper error reporting at the expense of a few more bytes.

Reply:
Yes, a fallible iterator is not very elegant, but `collect` and other combinators work on it.